drivers/i2c/i2c_ite_enhance.c

```diff
@@ -955,10 +958,99 @@ static int i2c_enhance_transfer(const struct device *dev,
 }
 
 #ifdef CONFIG_I2C_TARGET
-static void target_i2c_isr(const struct device *dev)
+static void target_i2c_isr_dma(const struct device *dev,
+			       uint8_t interrupt_status)
 {
 	struct i2c_enhance_data *data = dev->data;
+	const struct i2c_enhance_config *config = dev->config;
+	const struct i2c_target_callbacks *target_cb = data->target_cfg->callbacks;
 	struct i2c_target_dma_buffer *target_buffer = &data->target_buffer;
+	uint8_t *base = config->base;
+
+	/* Byte counter enable */
+	if (interrupt_status & IT8XXX2_I2C_IDW_CLR) {
+		IT8XXX2_I2C_BYTE_CNT_L(base) |=
+			(IT8XXX2_I2C_DMA_ADDR_RELOAD |
+			 IT8XXX2_I2C_BYTE_CNT_ENABLE);
+	}
+	/* The number of received data exceeds the byte counter setting */
+	if (interrupt_status & IT8XXX2_I2C_CNT_HOLD) {
+		LOG_ERR("The excess data written starts "
+			"from the memory address:%p",
+			target_buffer->in_buffer +
+			CONFIG_I2C_TARGET_IT8XXX2_MAX_BUF_SIZE);
+	}
+	/* Controller to write data */
+	if (interrupt_status & IT8XXX2_I2C_SLVDATAFLG) {
+		/* Number of receive data in target mode */
+		data->buffer_size =
+			((IT8XXX2_I2C_SLV_NUM_H(base) << 8) |
+			 IT8XXX2_I2C_SLV_NUM_L(base)) + 1;
+
+		/* Write data done callback function */
+		target_cb->buf_write_received(data->target_cfg,
+			target_buffer->in_buffer, data->buffer_size);
+	}
+	/* Controller to read data */
+	if (interrupt_status & IT8XXX2_I2C_IDR_CLR) {
+		uint32_t len;
+		uint8_t *rdata = NULL;
+
+		/* Clear byte counter setting */
+		IT8XXX2_I2C_BYTE_CNT_L(base) &=
+			~(IT8XXX2_I2C_DMA_ADDR_RELOAD |
+			  IT8XXX2_I2C_BYTE_CNT_ENABLE);
+		/* Read data callback function */
+		target_cb->buf_read_requested(data->target_cfg,
+			&rdata, &len);
+
+		if (len > CONFIG_I2C_TARGET_IT8XXX2_MAX_BUF_SIZE) {
+			LOG_ERR("The bufffer size exceeds "
+				"I2C_TARGET_IT8XXX2_MAX_BUF_SIZE: len=%d",
+				len);
+		} else {
+			memcpy(target_buffer->out_buffer, rdata, len);
+		}
+	}
+}
+
+static void target_i2c_isr_pio(const struct device *dev,
+			       uint8_t interrupt_status,
+			       uint8_t target_status)
+{
+	struct i2c_enhance_data *data = dev->data;
+	const struct i2c_enhance_config *config = dev->config;
+	const struct i2c_target_callbacks *target_cb = data->target_cfg->callbacks;
+	uint8_t *base = config->base;
+	uint8_t val;
+
+	/* Target ID write flag */
+	if (interrupt_status & IT8XXX2_I2C_IDW_CLR) {
+		target_cb->write_requested(data->target_cfg);
```

Comment:
The callback documentation indicates you should check the return value, and NACK if the callback returns an error.
https://github.com/zephyrproject-rtos/zephyr/blob/b42331476d3d88a16a3e82c9ac83ea7864fdfc61/include/zephyr/drivers/i2c.h#L279

You should check return values of the other callbacks as well.

Reply:
Done